biz.ganttproject.core/src/main/java/biz/ganttproject/core/chart/scene/gantt/TableSceneBuilder.kt

```diff
@@ -19,40 +19,104 @@ along with GanttProject.  If not, see <http://www.gnu.org/licenses/>.
 package biz.ganttproject.core.chart.scene.gantt
 
 import biz.ganttproject.core.chart.canvas.Canvas
+import biz.ganttproject.core.chart.canvas.TextMetrics
+import java.lang.Integer.max
 
 class TableSceneBuilder(private val config: Config) {
-  fun build(rows: List<Row>, canvas: Canvas = Canvas()): Canvas {
-    val dimensions = calculateDimensions(rows)
+  private lateinit var table: Table
+  private lateinit var canvas: Canvas
+  private lateinit var colsWidth: Map<Table.Column, Int>
+  private lateinit var dimensions: Dimension
+
+  fun build(table: Table, canvas: Canvas = Canvas()): Canvas {
+    this.table = table
+    this.canvas = canvas
+    colsWidth = calculateColsWidth(table)
+    dimensions = calculateDimensions(table, colsWidth)
     val state = PaintState(config.rowHeight)
 
-    rows.forEach {
+    paintHeader(state)
+    table.rows.forEach {
       val rectangle = canvas.createRectangle(
         0, state.y, dimensions.width, config.rowHeight
       )
       if (state.rowNumber % 2 == 1) {
         rectangle.style = "odd-row"
       }
-      paintRow(it, rectangle.middleY, canvas)
+      paintRow(it, rectangle.middleY)
       state.toNextRow()
     }
 
     return canvas
   }
 
-  private fun calculateDimensions(rows: List<Row>): Dimension {
-    val height = config.rowHeight * rows.size
-    val width = rows.map { it.width + it.indent + 2 * config.horizontalOffset }.maxOrNull() ?: 0
+  private fun calculateColsWidth(table: Table): Map<Table.Column, Int> {
+    val widths = mutableMapOf<Table.Column, Int>()
+    table.columns.forEach { col ->
+      val isFirst = col == table.columns.first()
+      widths[col] = if (col.width != null) {
+        col.width
+      } else {
+        val colNameWidth = config.textMetrics.getTextLength(col.name)
+        val colContentWidth = table.rows.map {
+          val indent = if (isFirst) it.indent else 0
+          indent + (it.values[col]?.run { config.textMetrics.getTextLength(this) } ?: 0)
+        }.maxOrNull() ?: 0
+        max(colNameWidth, colContentWidth)
+      }
+    }
+    return widths
+  }
+
+  private fun calculateDimensions(table: Table, colsWidth: Map<Table.Column, Int>): Dimension {
```

Comment:
`colsWidth` is passed here in the arguments while in `paintHeader` it is used as a membe variable. I think it is better to stick to a single way of using them.

Reply:
It was done so that it's harder to mess up with `lateinit`, but apparently we will get rid of it